src/Compilers/CSharp/Portable/Binder/ExpressionVariableFinder.cs

```diff
@@ -175,20 +177,61 @@ public override void VisitDeclarationExpression(DeclarationExpressionSyntax node
         {
             var context = node?.Parent  // ArgumentSyntax
                               ?.Parent  // ArgumentListSyntax
-                              ?.Parent; // invocation/constructor initializer
+                              ?.Parent; // invocation/constructor initializer/index access, etc.
             switch (context?.Kind())
             {
                 case SyntaxKind.InvocationExpression:
                 case SyntaxKind.ObjectCreationExpression:
+                case SyntaxKind.ElementAccessExpression:
+                    {
+                        var local = SourceLocalSymbol.MakeVariableDeclaredInExpression(
+                            _scopeBinder.ContainingMemberOrLambda, _scopeBinder, _enclosingBinderOpt, node.Type(),
+                            node.Identifier(), LocalDeclarationKind.RegularVariable, (ExpressionSyntax)context);
+                        _localsBuilder.Add(local);
+                        break;
+                    }
+
                 case SyntaxKind.ThisConstructorInitializer:
                 case SyntaxKind.BaseConstructorInitializer:
-                    var local = SourceLocalSymbol.MakeOutVariable(_scopeBinder.ContainingMemberOrLambda, _scopeBinder, _enclosingBinderOpt, node.Type(), node.Identifier(), context);
-                    _localsBuilder.Add(local);
-                    break;
+                    {
+                        Debug.Assert(_enclosingBinderOpt == null || _enclosingBinderOpt == _scopeBinder);
+                        var local = SourceLocalSymbol.MakeOutVariableInCtorInitializer(
+                            _scopeBinder.ContainingMemberOrLambda, _scopeBinder, node.Type(), node.Identifier(),
+                            (ConstructorInitializerSyntax)context);
+                        _localsBuilder.Add(local);
+                        break;
+                    }
+
+                case SyntaxKind.ElementBindingExpression:
+                    {
+                        context = context.Parent;
```

Comment:
>context = context.Parent [](start = 24, length = 24)

Thinking about conditional access scenarios in general, it feels like jumping out of just one level wouldn't be sufficient. I am leaning towards simplifying the logic and, for cases other than constructor initializers, always bubble-up to the top level expression within the statement. At the very least, we should jump out of all enclosing conditional accesses and this is true for other cases in this switch, other than those targeting constructor initializers.